File: Longest Substring Without Repeating Characters/mysolution3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        if(s.equals(""))
            return 0;
        
        Map<Character,Integer> map=new HashMap<>();
        
        int ans=0;
        
        int l=0,r=0;
        
        while(r<s.length())
        {
            char c=s.charAt(r);
            
            map.put(c,map.getOrDefault(c,0)+1);
            
            ans=Math.max(ans,map.size());
            // when repeating chars detected 
            if(map.size()<r-l+1)
            {   
                while(map.size()<r-l+1&&l<=r)
                {
                    c=s.charAt(l);
            
                    if(map.get(c)==1)
                        map.remove(c);
                    else
                        map.put(c,map.get(c)-1);
                    
                    l++;
                }
            }
            
            r++;
        }
        
        return ans;
    }
}

File: Longest Substring Without Repeating Characters/mysolution3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        set1=set([])
        n=len(s)
        
        start=0
        end=0
        ans=0
        
        while end<n:
            c=s[end]
            if c in set1:
                ans=max(ans,end-start)
                while c in set1:
                    set1.remove(s[start])
                    start+=1
                    
            set1.add(c)
            end+=1
        
        if len(set1)==(end-start):
            ans=max(ans,len(set1))
        
        return ans